superbench/benchmarks/docker_benchmarks/docker_base.py

```diff
@@ -36,24 +44,87 @@ def _preprocess(self):
+            logger.error('The container name is not set - benchmark: {}.'.format(self._name))
+            return False
+
+        output = run_command('docker pull {}'.format(self._image))
+        if output.returncode != 0:
+            self._result.set_return_code(ReturnCode.DOCKERBENCHMARK_IMAGE_PULL_FAILURE)
+            logger.error(
+                'Docker benchmark pull image failed - benchmark: {}, error message: {}.'.format(
+                    self._name, output.stdout
+                )
+            )
+            return False
+
+        return True
+
+    def _postprocess(self):
+        """Postprocess/cleanup operations after the benchmarking.
+
+        Return:
+            True if _postprocess() succeed.
+        """
+        rm_containers = 'docker rm $(docker stop {}))'.format(self._container)
+        run_command(rm_containers)
+
+        rm_image = 'docker rmi {}'.format(self._image)
+        run_command(rm_image)
+
+        return True
 
-    @abstractmethod
     def _benchmark(self):
-        """Implementation for benchmarking."""
-        pass
+        """Implementation for benchmarking.
 
-    def _process_raw_result(self, raw_output):
+        Return:
+            True if run benchmark successfully.
+        """
+        for cmd_idx in range(len(self._commands)):
+            logger.info(
+                'Execute command - round: {}, benchmark: {}, command: {}.'.format(
+                    self._curr_run_index, self._name, self._commands[cmd_idx]
+                )
+            )
+            output = run_command(self._commands[cmd_idx])
+            if output.returncode != 0:
+                self._result.set_return_code(ReturnCode.DOCKERBENCHMARK_EXECUTION_FAILURE)
+                logger.error(
+                    'Microbenchmark execution failed - round: {}, benchmark: {}, error message: {}.'.format(
+                        self._curr_run_index, self._name, output.stdout
+                    )
+                )
+                return False
+            else:
+                if not self._process_raw_result(cmd_idx, output.stdout):
+                    self._result.set_return_code(ReturnCode.DOCKERBENCHMARK_RESULT_PARSING_FAILURE)
+                    return False
+
+        return True
+
+    @abstractmethod
+    def _process_raw_result(self, cmd_idx, raw_output):
         """Function to process raw results and save the summarized results.
 
+          self._result.add_raw_data() and self._result.add_result() need to be called to save the results.
+
         Args:
-            raw_output (str): raw output string of the docker benchmark.
+            cmd_idx (int): the index of command corresponding with the raw_output.
+            raw_output (str): raw output string of the docker-benchmark.
 
         Return:
             True if the raw output string is valid and result can be extracted.
         """
-        # TODO: will implement it when add real benchmarks in the future.
```

Comment:
keep TODO?

Reply:
No need TODO now. This function will be realized by derived class.